File: Data Structures & Algorithms/climbing-stairs/submission-4.py
class Solution:
    def climbStairs(self, n: int) -> int:
        if n==1 : return 1
        map :Mapping[int,int]= {}
        def dfs(num)->int:
            if num > n:
                return 0
            if num == n:
                return 1
            if map.get(num):
                return map[num]
            map[num]= dfs(num+1) + dfs(num+2)
            return map[num]
        return dfs(0)

File: Data Structures & Algorithms/coin-change/submission-4.js
class Solution {
    /**
     * @param {number[]} coins
     * @param {number} amount
     * @return {number}
     */
    coinChange(coins, amount) {
        let  memo ={};
        coins.sort((a,b)=>b-a)

        function dfs(curAmount=0 , steps=0){
            if(memo[curAmount]!==undefined)return memo[curAmount];
            if(curAmount >amount)return -1;
            if(curAmount===amount)  return 0;
            let mn = Infinity
            for(let num of coins) {
                let ans = dfs(num+curAmount , steps+1)
                if(ans !== -1) mn = Math.min(mn , ans+1);
            }
            memo[curAmount] = mn
            return mn
        }
        let finalAns = dfs();
        return finalAns === Infinity ? -1 : finalAns;


    }
}
